File: latte/src/main/java/gg/beemo/latte/broker/BrokerClient.kt
package gg.beemo.latte.broker

import gg.beemo.latte.broker.rpc.BaseRpcRequestMessage
import gg.beemo.latte.broker.rpc.RpcClient
import gg.beemo.latte.broker.rpc.RpcResponse
import gg.beemo.latte.logging.Log
import kotlinx.coroutines.*
import java.util.Collections
import java.util.concurrent.atomic.AtomicBoolean


abstract class BrokerClient(
    @PublishedApi
    internal val connection: BrokerConnection,
    private val consumerScope: CoroutineScope = CoroutineScope(Dispatchers.Default + SupervisorJob()),
) {

    private val log by Log
    private val topics: MutableMap<String, TopicMetadata> = Collections.synchronizedMap(HashMap())

    inline fun <reified T> consumer(
        topic: String,
        key: String,
        options: BrokerClientOptions = BrokerClientOptions(),
        noinline callback: suspend CoroutineScope.(BaseBrokerMessage<T>) -> Unit,
    ): ConsumerSubclient<T> {
        return consumer(topic, key, options, T::class.java, isTypeNullable<T>(), callback)
    }

    @PublishedApi
    internal fun <T> consumer(
        topic: String,
        key: String,
        options: BrokerClientOptions = BrokerClientOptions(),
        type: Class<T>,
        isNullable: Boolean,
        callback: suspend CoroutineScope.(BaseBrokerMessage<T>) -> Unit,
    ): ConsumerSubclient<T> {
        log.debug("Creating consumer for key '{}' in topic '{}' with type {}", key, topic, type.name)
        return ConsumerSubclient(connection, this, topic, key, options, type, isNullable, callback).also {
            registerSubclient(it)
        }
    }

    inline fun <reified T> producer(
        topic: String,
        key: String,
        options: BrokerClientOptions = BrokerClientOptions(),
    ): ProducerSubclient<T> {
        return producer(topic, key, options, T::class.java, isTypeNullable<T>())
    }

    @PublishedApi
    internal fun <T> producer(
        topic: String,
        key: String,
        options: BrokerClientOptions = BrokerClientOptions(),
        type: Class<T>,
        isNullable: Boolean,
    ): ProducerSubclient<T> {
        log.debug("Creating producer for key '{}' in topic '{}' with type {}", key, topic, type.name)
        return ProducerSubclient(connection, this, topic, key, options, type, isNullable).also {
            registerSubclient(it)
        }
    }

    inline fun <reified RequestT, reified ResponseT> rpc(
        topic: String,
        key: String,
        options: BrokerClientOptions = BrokerClientOptions(),
        noinline callback: suspend CoroutineScope.(BaseRpcRequestMessage<RequestT, ResponseT>) -> RpcResponse<ResponseT>,
    ): RpcClient<RequestT, ResponseT> {
        return RpcClient(
            this,
            topic,
            key,
            options,
            RequestT::class.java,
            isTypeNullable<RequestT>(),
            ResponseT::class.java,
            isTypeNullable<ResponseT>(),
            callback,
        )
    }

    private fun registerSubclient(subclient: BaseSubclient) {
        val topic = subclient.topic
        val metadata = topics.computeIfAbsent(topic) {
            TopicMetadata(connection, consumerScope, topic)
        }
        metadata.registerSubclient(subclient)
    }

    internal fun deregisterSubclient(subclient: BaseSubclient) {
        val topic = subclient.topic
        topics[topic]?.let {
            it.deregisterSubclient(subclient)
            if (it.isEmpty) {
                it.destroy()
                topics.remove(topic)
            }
        }
    }

    fun destroy(cancelScope: Boolean = true) {
        log.debug("Destroying BrokerClient of type {} with active topics {}", javaClass.simpleName, topics.keys)
        while (topics.isNotEmpty()) {
            val topic = topics.keys.first()
            topics[topic]?.destroy()
            topics.remove(topic)
        }
        if (cancelScope) {
            consumerScope.cancel()
        }
    }

}

private class TopicMetadata(
    private val connection: BrokerConnection,
    private val consumerScope: CoroutineScope,
    private val topic: String,
) {

    private val log by Log
    private val _keys: MutableMap<String, KeyMetadata> = Collections.synchronizedMap(HashMap())
    private val isBeingDestroyed = AtomicBoolean(false)
    val isEmpty: Boolean
        get() = _keys.isEmpty()

    @Volatile
    private var connectionListener: TopicListener? = null

    fun registerSubclient(subclient: BaseSubclient) {
        log.debug(
            "Adding {} for key '{}' in topic '{}'",
            subclient.javaClass.simpleName,
            subclient.key,
            subclient.topic
        )
        check(subclient.topic == topic) {
            "Attempting to register subclient with topic '${subclient.topic}' in TopicMetadata of '$topic'"
        }
        val metadata = getOrCreateKeyMetadata(subclient.key)
        when (subclient) {
            is ConsumerSubclient<*> -> {
                if (metadata.consumers.isEmpty() && connectionListener == null) {
                    log.debug("Creating new connection listener for topic '{}'", subclient.topic)
                    // New consumer - create a new connection listener for this topic
                    val listener = TopicListener { topic, key, value, headers ->
                        onTopicMessage(topic, key, value, headers)
                    }
                    connection.on(subclient.topic, listener)
                    connectionListener = listener
                }
                metadata.consumers.add(subclient)
            }

            is ProducerSubclient<*> -> {
                metadata.producers.add(subclient)
            }
        }
    }

    fun deregisterSubclient(subclient: BaseSubclient) {
        log.debug(
            "Removing {} for key '{}' in topic '{}'",
            subclient.javaClass.simpleName,
            subclient.key,
            subclient.topic
        )
        val metadata = getExistingKeyMetadata(subclient.key)
        metadata?.let {
            when (subclient) {
                is ConsumerSubclient<*> -> it.consumers.remove(subclient)
                is ProducerSubclient<*> -> it.producers.remove(subclient)
            }
            maybeCleanupKeyMetadata(it)
        }
    }

    private fun maybeCleanupKeyMetadata(keyMetadata: KeyMetadata) {
        if (keyMetadata.isEmpty) {
            _keys.remove(keyMetadata.key)
        }
        if (this.isEmpty) {
            connectionListener?.let {
                log.debug("Removing connection listener for topic '{}' after key cleanup", topic)
                connection.off(topic, it)
                connectionListener = null
            }
        }
    }

    private fun getOrCreateKeyMetadata(key: String): KeyMetadata {
        return _keys.computeIfAbsent(key) {
            KeyMetadata(key)
        }
    }

    private fun getExistingKeyMetadata(key: String): KeyMetadata? {
        return _keys[key]
    }

    private fun onTopicMessage(
        topic: String,
        key: String,
        value: String,
        headers: BrokerMessageHeaders,
    ) {
        val metadata = getExistingKeyMetadata(key) ?: return
        for (consumer in metadata.consumers) {
            consumerScope.launch {
                try {
                    consumer.onIncomingMessage(value, headers)
                } catch (e: Exception) {
                    log.error("Uncaught error in BrokerClient listener for key '$key' in topic '$topic'", e)
                }
            }
        }
    }

    fun destroy() {
        if (!isBeingDestroyed.compareAndSet(false, true)) {
            return
        }
        while (_keys.isNotEmpty()) {
            val key = _keys.keys.first()
            _keys[key]?.destroy()
            _keys.remove(key)
        }
        connectionListener?.let {
            log.debug("Removing connection listener for topic '{}' in destroy()", topic)
            connection.off(topic, it)
            connectionListener = null
        }
    }

}

private class KeyMetadata(val key: String) {
    val producers: MutableSet<ProducerSubclient<*>> = Collections.synchronizedSet(HashSet())
    val consumers: MutableSet<ConsumerSubclient<*>> = Collections.synchronizedSet(HashSet())

    val isEmpty: Boolean
        get() = producers.isEmpty() && consumers.isEmpty()

    fun destroy() {
        producers.forEach(ProducerSubclient<*>::destroy)
        consumers.forEach(ConsumerSubclient<*>::destroy)
        producers.clear()
        consumers.clear()
    }
}

@PublishedApi
internal inline fun <reified T> isTypeNullable(): Boolean {
    return null is T || T::class.java == Unit::class.java || T::class.java == Void::class.java
}

File: water/src/broker/BrokerClient.ts
import type { BaseIssue, BaseSchema, InferOutput } from "valibot";
import type {
	BaseSubclient,
	BrokerConnection,
	BrokerMessage,
	BrokerMessageHeaders,
	RpcRequestMessage,
	RpcResponse,
	TopicListener,
} from "@";
import { BrokerClientOptions, ConsumerSubclient, Logger, ProducerSubclient, RpcClient } from "@";

const TAG = "BrokerClient";

export abstract class BrokerClient {
	private readonly topics: Map<string, TopicMetadata> = new Map();

	public constructor(public readonly connection: BrokerConnection) {}

	public consumer<TSchema extends BaseSchema<unknown, unknown, BaseIssue<unknown>>>(
		topic: string,
		key: string,
		schema: TSchema,
		callback: (msg: BrokerMessage<InferOutput<TSchema>>) => Promise<void>,
		options: BrokerClientOptions = new BrokerClientOptions(),
	): ConsumerSubclient<TSchema> {
		Logger.debug(TAG, `Creating consumer for key '${key}' in topic '${topic}'`);
		const client = new ConsumerSubclient(this.connection, this, topic, key, options, schema, callback);
		this.registerSubclient(client);
		return client;
	}

	public producer<T>(
		topic: string,
		key: string,
		options: BrokerClientOptions = new BrokerClientOptions(),
	): ProducerSubclient<T> {
		Logger.debug(TAG, `Creating producer for key '${key}' in topic '${topic}'`);
		const client = new ProducerSubclient(this.connection, this, topic, key, options);
		this.registerSubclient(client);
		return client;
	}

	public rpc<
		RequestTSchema extends BaseSchema<unknown, unknown, BaseIssue<unknown>>,
		ResponseTSchema extends BaseSchema<unknown, unknown, BaseIssue<unknown>>,
	>(
		topic: string,
		key: string,
		requestSchema: RequestTSchema,
		responseSchema: ResponseTSchema,
		callback: (
			msg: RpcRequestMessage<InferOutput<RequestTSchema>, InferOutput<ResponseTSchema>>,
		) => Promise<RpcResponse<InferOutput<ResponseTSchema>>>,
		options: BrokerClientOptions = new BrokerClientOptions(),
	): RpcClient<RequestTSchema, ResponseTSchema> {
		return new RpcClient(this, topic, key, options, requestSchema, responseSchema, callback);
	}

	private registerSubclient(subclient: BaseSubclient): void {
		const topic = subclient.topic;
		if (!this.topics.has(topic)) {
			this.topics.set(topic, new TopicMetadata(this.connection, topic));
		}
		const metadata = this.topics.get(topic)!;
		metadata.registerSubclient(subclient);
	}

	// Private API
	public deregisterSubclient(subclient: BaseSubclient): void {
		const topic = subclient.topic;
		if (this.topics.has(topic)) {
			const metadata = this.topics.get(topic)!;
			metadata.deregisterSubclient(subclient);
			if (metadata.isEmpty) {
				this.topics.delete(topic);
			}
		}
	}

	public destroy(): void {
		Logger.debug(TAG, `Destroying BrokerClient with active topics: ${[...this.topics.keys()]}`);
		while (this.topics.size > 0) {
			const [topic, metadata] = this.topics.entries().next().value as [string, TopicMetadata];
			metadata.destroy();
			this.topics.delete(topic);
		}
	}
}

class TopicMetadata {
	private static readonly TAG = "TopicMetadata";

	private readonly keys: Map<string, KeyMetadata> = new Map();
	private isDestroyed: boolean = false;
	private connectionListener: TopicListener | null = null;

	public constructor(
		private readonly connection: BrokerConnection,
		private readonly topic: string,
	) {}

	public get isEmpty(): boolean {
		return this.keys.size === 0;
	}

	public registerSubclient(subclient: BaseSubclient): void {
		if (subclient.topic !== this.topic) {
			throw new Error(
				`Attempting to register subclient with topic '${subclient.topic}' in TopicMetadata of '${this.topic}'`,
			);
		}
		Logger.debug(
			TopicMetadata.TAG,
			`Adding ${this.constructor.name} for key '${subclient.key}' in topic '${this.topic}'`,
		);
		const metadata = this.getOrCreateKeyMetadata(subclient.key);
		if (subclient instanceof ConsumerSubclient) {
			if (metadata.consumers.size === 0 && this.connectionListener === null) {
				Logger.debug(TopicMetadata.TAG, `Creating new connection listener for topic '${this.topic}'`);
				this.connectionListener = (topic, key, value, headers) => this.onTopicMessage(topic, key, value, headers);
				this.connection.on(this.topic, this.connectionListener);
			}
			metadata.consumers.add(subclient);
		} else if (subclient instanceof ProducerSubclient) {
			metadata.producers.add(subclient);
		}
	}

	public deregisterSubclient(subclient: BaseSubclient): void {
		Logger.debug(
			TopicMetadata.TAG,
			`Removing ${this.constructor.name} for key '${subclient.key}' in topic '${this.topic}'`,
		);
		const metadata = this.getExistingKeyMetadata(subclient.key);
		if (metadata) {
			metadata.producers.delete(subclient as ProducerSubclient<any>);
			metadata.consumers.delete(subclient as ConsumerSubclient<any>);
			this.maybeCleanupKeyMetadata(metadata);
		}
	}

	private maybeCleanupKeyMetadata(metadata: KeyMetadata): void {
		if (metadata.isEmpty) {
			this.keys.delete(metadata.key);
		}
		if (this.isEmpty && this.connectionListener !== null) {
			Logger.debug(TopicMetadata.TAG, `Removing connection listener for topic '${this.topic}' after key cleanup`);
			this.connection.off(this.topic, this.connectionListener);
			this.connectionListener = null;
		}
	}

	private getOrCreateKeyMetadata(key: string): KeyMetadata {
		if (!this.keys.has(key)) {
			this.keys.set(key, new KeyMetadata(key));
		}
		return this.keys.get(key)!;
	}

	private getExistingKeyMetadata(key: string): KeyMetadata | null {
		return this.keys.get(key) ?? null;
	}

	private onTopicMessage(topic: string, key: string, value: string, headers: BrokerMessageHeaders): void {
		const metadata = this.getExistingKeyMetadata(key);
		if (!metadata) {
			return;
		}
		for (const consumer of metadata.consumers) {
			consumer
				.onIncomingMessage(value, headers)
				.catch(e =>
					Logger.error(
						TopicMetadata.TAG,
						`Uncaught error in BrokerClient listener for key '${key}' in topic '${topic}'`,
						e,
					),
				);
		}
	}

	public destroy(): void {
		if (this.isDestroyed) {
			return;
		}
		while (this.keys.size > 0) {
			const [key, metadata] = this.keys.entries().next().value as [string, KeyMetadata];
			metadata.destroy();
			this.keys.delete(key);
		}
		if (this.connectionListener !== null) {
			Logger.debug(TopicMetadata.TAG, `Removing connection listener for topic '${this.topic}' during destroy`);
			this.connection.off(this.topic, this.connectionListener);
			this.connectionListener = null;
		}
	}
}

class KeyMetadata {
	public readonly producers: Set<ProducerSubclient<any>> = new Set();
	public readonly consumers: Set<ConsumerSubclient<any>> = new Set();

	public constructor(public readonly key: string) {}

	public get isEmpty(): boolean {
		return this.producers.size === 0 && this.consumers.size === 0;
	}

	public destroy(): void {
		this.producers.forEach(producer => producer.destroy());
		this.consumers.forEach(consumer => consumer.destroy());
		this.producers.clear();
		this.consumers.clear();
	}
}
